File: Dynamic_Programming/CanSum.py
"""
write a function 'canSum(targetSum, numbers)' that takes in a targetSum
and an array of numbers as arguments.
The function should return a boolean indicating whether it is possible to
generate the targetSum suing numbers from the array.

You may use an element of the array as many times as needed.
you may assume that all input numbers are nonnegative.
"""

def canSum(n, intList):
    if (n == 0):
        return True
    if (n < 0):
        return False

    for int in intList:
        remainder = n - int
        if (canSum(remainder, intList)):
            return True

    return False;


# Dynamic programming using memoization:
def canSumDP(n, intList, table):
    if n in table:
        return table[n]
    if (n == 0):
        return True
    if (n < 0):
        return False

    for int in intList:
        remainder = n - int
        table[remainder] = canSumDP(remainder, intList, table)
        if (table[remainder]):
            return True

    return False


def canSum_tabulation(targetSum, numbers):
    table = [False] * (targetSum + 1)
    table[0] = True

    for num in range(targetSum + 1):
        if (table[num]):
            for int in numbers:
                if (num + int <= targetSum):
                    table[num + int] = True

    return table[targetSum]




print(canSum(7, [3, 4]))
print(canSum(7, [5, 3]))
# print(canSum(300, [7, 14]))
print(canSumDP(300, [7, 14], dict()))
print(canSum_tabulation(300, [7, 14]))
print(canSum_tabulation(21, [5, 2]))

File: Dynamic_Programming/CanSum.java
/*
    write a function 'canSum(targetSum, numbers)' that takes in a targetSum
    and an array of numbers as arguments.
    The function should return a boolean indicating whether it is possible to
    generate the targetSum suing numbers from the array.

    You may use an element of the array as many times as needed.
    you may assume that all input numbers are nonnegative.

    canSum(7, [5, 3, 4, 7]) --> true because: 3 + 4 = 7 or 7 == 7
    canSum(7, [2, 4]) --> false
 */
import java.util.Map;
import java.util.HashMap;

public class CanSum {

    public boolean canSum(int targetSum, int[] numbers) {
        if (targetSum == 0) return true;
        if (targetSum < 0) return false;

        for (int i = 0; i < numbers.length; i += 1) {
            int remainder = targetSum - numbers[i];
            if (canSum(remainder, numbers)) return true;
        }
        return false;
    }

    //Dynamic programming using memoization:
    public boolean canSum(int targetSum, int[] numbers, Map<Integer, Boolean> table) {
        if (table.containsKey(targetSum)) return table.get(targetSum);
        if (targetSum == 0) return true;
        if (targetSum < 0) return false;

        for (int i = 0; i < numbers.length; i += 1) {
            int remainder = targetSum - numbers[i];
            table.put(remainder, canSum(remainder, numbers, table));
            if (table.get(remainder)) return true;
        }

        return false;
    }

    public boolean canSumTabulation(int targetSum, int[] numbers) {
        boolean[] table = new boolean[targetSum + 1];
        table[0] = true;

        for (int i = 0; i <= targetSum; i += 1) {
            //If I can get to current amount
            if (table[i]) {
                for (int num : numbers) {
                    if (i + num <= targetSum) {
                        table[i + num] = true;
                    }
                }
            }
        }

        return table[targetSum];
    }


    public static void main(String[] agrs) {
        CanSum cs = new CanSum();
        System.out.println(cs.canSum(7, new int[]{3, 4}));
        System.out.println(cs.canSum(7, new int[]{5, 3}));
        //System.out.println(cs.canSum(300, new int[]{7, 14}));
        System.out.println(cs.canSum(300, new int[]{7, 14}, new HashMap<Integer, Boolean>()));
        System.out.println(cs.canSumTabulation(300, new int[]{7, 14}));
        System.out.println(cs.canSumTabulation(30, new int[]{7, 14, 10}));
    }
}
